ql/src/java/org/apache/hadoop/hive/ql/ppd/OpProcFactory.java

```diff
@@ -782,6 +790,89 @@ protected ExprWalkerInfo mergeChildrenPred(Node nd, OpWalkerInfo owi,
     }
   }
 
+  protected static Object splitFilter(FilterOperator op,
+      ExprWalkerInfo ewi, OpWalkerInfo owi) throws SemanticException {
+
+    RowSchema inputRS = op.getSchema();
+
+    Map<String, List<ExprNodeDesc>> pushDownPreds = ewi.getFinalCandidates();
+    Map<String, List<ExprNodeDesc>> unPushDownPreds = ewi.getNonFinalCandidates();
+
+    // combine all deterministic predicates into a single expression
+    List<ExprNodeDesc> deterministicPreds = new ArrayList<ExprNodeDesc>();
+    Iterator<List<ExprNodeDesc>> iterator1 = pushDownPreds.values().iterator();
+    while (iterator1.hasNext()) {
+      for (ExprNodeDesc pred : iterator1.next()) {
+        deterministicPreds = ExprNodeDescUtils.split(pred, deterministicPreds);
+      }
+    }
+
+    if (deterministicPreds.isEmpty()) {
+      return null;
+    }
+
+    List<ExprNodeDesc> nondeterministicPreds = new ArrayList<ExprNodeDesc>();
+    Iterator<List<ExprNodeDesc>> iterator2 = unPushDownPreds.values().iterator();
+    while (iterator2.hasNext()) {
+      for (ExprNodeDesc pred : iterator2.next()) {
+        nondeterministicPreds = ExprNodeDescUtils.split(pred, nondeterministicPreds);
+      }
+    }
+
+    assert !nondeterministicPreds.isEmpty();
+
+    ExprNodeDesc deterministicCondn = ExprNodeDescUtils.mergePredicates(deterministicPreds);
+    ExprNodeDesc nondeterministicCondn = ExprNodeDescUtils.mergePredicates(nondeterministicPreds);
+
+    Operator<FilterDesc> deterministicFilter =
+        OperatorFactory.get(new FilterDesc(deterministicCondn, false), new RowSchema(inputRS.getSignature()));
+
+    deterministicFilter.setChildOperators(new ArrayList<Operator<? extends OperatorDesc>>());
```

Comment:
use OpProcFactory.createFilter may be better?

Reply:
oh, it's bad to use createFilter for createFilter add filter which is current operator's child，however, we need a filter which is current operator's parent